feat: add page for best practices in using system (#366)
openfga/openfga.dev: docs/content/getting-started/tuples-api-best-practices.mdx

---
title: Managing Tuples and Invoking API Best Practices
description: Best Practices of Managing Tuples and Invoking APIs
sidebar_position: 8
slug: /getting-started/tuples-api-best-practices
---

import {
  DocumentationNotice,
  ProductName,
  ProductNameFormat,
  RelatedSection,
} from "@components/Docs";

# Best Practices of Managing Tuples and Invoking APIs

<DocumentationNotice />

The following list outlines some guidelines and best-practices for using <ProductName format={ProductNameFormat.ShortForm}/>:

- Do not store Personal Identifiable Information in tuples
- Always specify authorization model ID whenever possible
- Avoid using intersection or exclusion in ListObjects authorization model

## Do Not Store Personal Identifiable Information in Tuples

> You can use any string for user and object identifiers, but we strongly recommend that you do not store Personal Identifiable Information (PII) tuples for the following reasons:

Reviewer: Can you add a disclaimer before or after this section that the docs does use readable identifiers to make it easier to follow?

Author: done

1. By not storing personal information, you will reduce the impact of using the product from a compliance perspective.

> 2. In some cases, tuple information flows through Data Sub Processors such as ones used for logging and tracing. You’ll minimize risk if you don’t include sensitive data in tuples.

Author: done

> 2. In some cases, tuple information flows through Data Sub Processors such as ones used for logging and tracing. You’ll minimize risk if you don’t include sensitive data in tuples.

Author: updated

## Always Specify Authorization Model ID Whenever Possible

> It is strong recommended that authorization model ID be specified in your Relationship Queries (such as [Check](./perform-check.mdx) and [ListObjects](../interacting/relationship-queries.mdx#listobjects)) and Relationship Tuples (such as [Read](../interacting/relationship-queries.mdx#read) and [Write](./update-tuples.mdx)) request.

Author: read does not allow auth id

Specifying authorization model ID in API calls have the following advantages:
1. Better performances as <ProductName format={ProductNameFormat.ShortForm}/> do not need to invoke database queries to get the latest authorization model ID 
2. Allows consistent behavior in your production system until until you are ready to switch to the new model.

> ## Avoid Using Intersection or Exclusion in ListObjects Authorization Model

Author: updated

---
title: Managing Tuples and Invoking API Best Practices
description: Best Practices of Managing Tuples and Invoking APIs
sidebar_position: 8
slug: /getting-started/tuples-api-best-practices
---

import {
  DocumentationNotice,
  ProductName,
  ProductNameFormat,
  RelatedSection,
} from "@components/Docs";

# Best Practices of Managing Tuples and Invoking APIs

<DocumentationNotice />

> The following list outlines some guidelines and best-practices for using <ProductName format={ProductNameFormat.ShortForm}/>:

Reviewer 2: ```suggestion
The following list outlines some guidelines and best practices for using <ProductName format={ProductNameFormat.ShortForm}/>:
```

- Do not store Personal Identifiable Information in tuples
- Always specify authorization model ID whenever possible
- If using ListObjects, avoid having intersection or exclusion in the authorization model

## Do Not Store Personal Identifiable Information in Tuples

You can use any string for user and object identifiers, but we strongly recommend that you do not store Personal Identifiable Information (PII) tuples for the following reasons:

1. By not storing personal information, you will reduce the impact of using the product from a compliance perspective.
2. In some cases, tuple information might flow through Data Sub Processors such as ones used for logging and tracing. You’ll minimize risk if you don’t include sensitive data in tuples.
3. You will ensure uniqueness. For example, different users may have the same user name. Using unique opaque identifiers removes the risk of adding incorrect relationships.

:::caution Note
The documentation and samples uses first names and simple ids to illustrate easy-to-follow examples.
:::

## Always Specify Authorization Model ID Whenever Possible

> It is strong recommended that authorization model ID be specified in your Relationship Queries (such as [Check](./perform-check.mdx) and [ListObjects](../interacting/relationship-queries.mdx#listobjects)) and Relationship Tuples (such as [Write](./update-tuples.mdx)) request.

Reviewer 2: ```suggestion
It is strongly recommended that authorization model ID be specified in your Relationship Queries (such as [Check](./perform-check.mdx) and [ListObjects](../interacting/relationship-queries.mdx#listobjects)) and Relationship Tuples Commands (such as [Write](./update-tuples.mdx)).
```

Specifying authorization model ID in API calls have the following advantages:

> 1. Better performances as <ProductName format={ProductNameFormat.ShortForm}/> do not need to invoke database queries to get the latest authorization model ID 

Reviewer 2: ```suggestion
1. Better performance, as <ProductName format={ProductNameFormat.ShortForm}/> will not need to perform a database query to get the latest authorization model ID 
2. ```

- Do not store Personal Identifiable Information in tuples
- Always specify authorization model ID whenever possible
- If using ListObjects, avoid having intersection or exclusion in the authorization model

## Do Not Store Personal Identifiable Information in Tuples

You can use any string for user and object identifiers, but we strongly recommend that you do not store Personal Identifiable Information (PII) tuples for the following reasons:

1. By not storing personal information, you will reduce the impact of using the product from a compliance perspective.
2. In some cases, tuple information might flow through Data Sub Processors such as ones used for logging and tracing. You’ll minimize risk if you don’t include sensitive data in tuples.

> 3. You will ensure uniqueness. For example, different user may have the same user name. Using unique opaque identifiers removes the risk of adding incorrect relationships.

Reviewer: ```suggestion
3. You will ensure uniqueness. For example, different users may have the same user name. Using unique opaque identifiers removes the risk of adding incorrect relationships.
```

:::caution Note

> This document uses first names and simple ids to illustrate easy-to-follow examples.

Reviewer: ```suggestion
The documentation and samples uses first names and simple ids to illustrate easy-to-follow examples.
```
